src/server/editorServices.ts

```diff
@@ -463,17 +466,33 @@ namespace ts.server {
             project.updateGraph();
         }
 
-        setCompilerOptionsForInferredProjects(projectCompilerOptions: protocol.ExternalProjectCompilerOptions): void {
-            this.compilerOptionsForInferredProjects = convertCompilerOptions(projectCompilerOptions);
+        setCompilerOptionsForInferredProjects(projectCompilerOptions: protocol.ExternalProjectCompilerOptions, projectRootPath?: string): void {
+            // ignore this settings if we are not creating inferred projects per project root.
+            if (projectRootPath && !this.useInferredProjectPerProjectRoot) return;
+
+            const compilerOptionsForInferredProjects = convertCompilerOptions(projectCompilerOptions);
+
             // always set 'allowNonTsExtensions' for inferred projects since user cannot configure it from the outside
             // previously we did not expose a way for user to change these settings and this option was enabled by default
-            this.compilerOptionsForInferredProjects.allowNonTsExtensions = true;
-            this.compileOnSaveForInferredProjects = projectCompilerOptions.compileOnSave;
-            for (const proj of this.inferredProjects) {
-                proj.setCompilerOptions(this.compilerOptionsForInferredProjects);
-                proj.compileOnSaveEnabled = projectCompilerOptions.compileOnSave;
+            compilerOptionsForInferredProjects.allowNonTsExtensions = true;
+
+            if (projectRootPath) {
+                this.compilerOptionsForInferredProjectsPerProjectRoot.set(projectRootPath, compilerOptionsForInferredProjects);
+            }
+            else {
+                this.compilerOptionsForInferredProjects = compilerOptionsForInferredProjects;
             }
-            this.updateProjectGraphs(this.inferredProjects);
+
+            const updatedProjects: Project[] = [];
+            for (const project of this.inferredProjects) {
+                if (project.projectRootPath === projectRootPath || (project.projectRootPath && !this.compilerOptionsForInferredProjectsPerProjectRoot.has(project.projectRootPath))) {

```

Comment:
Don't understand the second half of this condition -- if the project has a root path set but it's somehow not in the map, update it anyway even if its root path doesn't match?